### PR TITLE
Refactor/ddf2

### PR DESCRIPTION
**Pull request by zkidkid**
Please DON'T MERGE. This PR is only for code review.

Change for Schema & IDataSource.
Basically all ddf will be created from an IDataSource
In case of engine doesn't support a concrete datasource -> throw unsupported datasource exception
else Engine will do prepare and make ddf on this datasource.

/cc @sameerp @mbbui @steveyan777 @Huandao0812 @nhanitvn @hai-adatao @lebinh

### CODE REVIEW

**Comment by zkidkid** [comment]
@sameerp @PangZhi @mbbui @nhanitvn @Huandao0812 @lebinh @phvu could you help to review & comment on it.
more doc: https://docs.google.com/document/d/1iTAUREZnfua1xMEvjLpC4CO69zaXI-74QANRDqZnn9s/edit

File: core/src/main/java/io/ddf/content/Schema.java [hunk 1]
```diff
@@ -1,655 +1,193 @@
...
+    columns = new ArrayList<>();
   }
 
-  public String getTableName() {
-    return mTableName;
+  public Schema(Schema schema) {
+    this();
+    this.colNames.addAll(schema.colNames);
+    columns.addAll(schema.getColumns());
+    assert colNames.size() == columns.size();
```
Comment (sameerp) [inline core/src/main/java/io/ddf/content/Schema.java]: In general, don't use `assert` to do argument checking. Use Guava `Preconditions` or something equivalent.
See this page on [assertions] (https://docs.oracle.com/javase/8/docs/technotes/guides/language/assert.html) for more details

> Do not use assertions for argument checking in public methods.
Argument checking is typically part of the published specifications (or contract) of a method, and these specifications must be obeyed whether assertions are enabled or disabled.

File: core/src/main/java/io/ddf/content/Schema.java [hunk 2]
```diff
@@ -1,655 +1,193 @@
...
-      return null;
-    String[] segments = columnList.split("\\s*,\\s*");
+@NotThreadSafe
+public  class Schema {
 
-    mColumns.clear();
-    for (String segment : segments) {
-      if (Strings.isNullOrEmpty(segment))
-        continue;
+  protected List<String> colNames;
```
Comment (sameerp) [inline core/src/main/java/io/ddf/content/Schema.java]: Why are these `protected`? Do we want subclasses of `Schema`?

Comment (sameerp) [inline core/src/main/java/io/ddf/content/Schema.java]: Why is there are separate list of column names and columns? Why not just one list of columns?

File: core/src/main/java/io/ddf/content/Schema.java [hunk 3]
```diff
@@ -1,655 +1,193 @@
...
-  public Column getColumn(int i) {
-    if (mColumns.isEmpty()) {
-      return null;
-    }
-    if (i < 0 || i >= mColumns.size()) {
-      return null;
-    }
-
-    return mColumns.get(i);
+  public int getColumnIndex(String columnName) throws DDFException {
```
Comment (sameerp) [inline core/src/main/java/io/ddf/content/Schema.java]: This does not throw `DDFException`.

File: core/src/main/java/io/ddf/content/Schema.java [hunk 4]
```diff
@@ -1,655 +1,193 @@
...
   public String toString() {
-    Iterable<String> columnSpecs = Iterables.transform(getColumns(), new Function<Column, String>() {
-      @Override
-      public String apply(Column column) {
-        return column.getName() + " " + column.getType();
+    StringBuilder sb = new StringBuilder();
+    for(Column column : columns){
+      try {
+        sb.append(column.getName() + ":" + column.getType().getSimpleName() + "|");
+      }catch(Exception ex){
```
Comment (sameerp) [inline core/src/main/java/io/ddf/content/Schema.java]: What exception do you expect here?

File: core/src/main/java/io/ddf/content/Schema.java [hunk 5]
```diff
@@ -1,655 +1,193 @@
...
-          temp = new HashMap<String, java.lang.Double>();
-          while (valuesIterator.hasNext()) {
-            String columnValue = valuesIterator.next();
-            temp.put(columnValue, Double.parseDouble(i + ""));
-            i += 1;
-          }
-          dummyCoding.getMapping().put(currentColumnIndex, temp);
-          count += temp.size() - 1;
+      @Override
+      protected Class inferType(String typeName) throws SchemaException {
```
Comment (sameerp) [inline core/src/main/java/io/ddf/content/Schema.java]: Are these the only types we can support at the DDF level? Spark RDDs can contain other arbitrary types. This appears like a departure from our existing implementation.

File: core/src/main/java/io/ddf/content/Schema.java [hunk 6]
```diff
@@ -1,655 +1,193 @@
...
-    this.mTableName = mTableName;
+  public int getNumColumn() {
+    return columns.size();
   }
 
   public List<Column> getColumns() {
-    List<Column> columns = new ArrayList<Column>();
-    for (Column column : mColumns) {
-      columns.add(column.clone());
-    }
```
Comment (Huandao0812) [inline core/src/main/java/io/ddf/content/Schema.java]: getColumns should clone the list of columns and return the cloned list, avoiding user to modify the schema by modifying the columns

**Pull request closed without merging**